pkg/agent/core/ngt/handler/grpc/index_test.go

```diff
@@ -1029,106 +1029,877 @@ func Test_server_SaveIndex(t *testing.T) {
+				return err
 			}
-			if test.afterFunc != nil {
-				defer test.afterFunc(test.args)
+			if !reflect.DeepEqual(obj, ir.GetVector()) {
+				return errors.Errorf("vector is not match, got: %v, want: %v", obj, ir)
 			}
-			checkFunc := test.checkFunc
-			if test.checkFunc == nil {
-				checkFunc = defaultCheckFunc
+		}
+
+		// check total index count is same
+		ii, err := srv.IndexInfo(ctx, emptyPayload)
+		if err != nil {
+			return err
+		}
+
+		wantIndexInfo := &payload.Info_Index_Count{
+			Stored: uint32(len(wantVecs)),
+		}
+		if !reflect.DeepEqual(ii, wantIndexInfo) {
+			return errors.Errorf("stored index count not correct, got: %v, want: %v", ii, wantIndexInfo)
+		}
+
+		return nil
+	}
+
+	/*
+		- Equivalence Class Testing (with copy on write disable)
+			- case 1.1: success to create and save 1 uncommitted insert index
+			- case 1.2: success to create and save 100 uncommitted insert index
+			- case 2.1: success to create and save 1 uncommitted delete index
+			- case 2.2: success to create and save 100 uncommitted delete index
+			- case 3.1: success to create and save 1 uncommitted update index
+			- case 3.2: success to create and save 100 uncommitted update index
+		- Boundary Value Testing
+			- case 1.1: fail to create and save 0 index
+			- case 2.1: success to create and save index with invalid dimension
+				- the invalid index will be removed from NGT and the index file
+		- Decision Table Testing
+			- case 1.1: success to create and save 100 index with in-memory mode
+				- do nothing and no file will be created
+			- case 2.1: success to create and save 1 inserted index with copy-on-write enabled
+			- case 2.2: success to create and save 100 inserted index with copy-on-write enabled
+
+			// with uncommitted index count 100
+			- case 3.1: success to create and save index with poolSize > uncommitted index count
+			- case 3.2: success to create and save index with poolSize < uncommitted index count
+			- case 3.3: success to create and save index with poolSize = uncommitted index count
+			- case 3.4: success to create and save index with poolSize = 0
+	*/
+	tests := []test{
+		func() test {
+			insertCnt := 1
+			var ir *payload.Insert_MultiRequest
+
+			return test{
+				name: "Equivalence Class Testing case 1.1: success to create and save 1 uncommitted insert index",
+				args: args{
+					c: &payload.Control_CreateIndexRequest{
+						PoolSize: uint32(insertCnt),
+					},
+				},
+				fields: fields{
+					srvOpts:   defaultSrvOpts,
+					svcCfg:    defaultSvcCfg,
+					svcOpts:   defaultSvcOpts,
+					indexPath: mkdirTemp(),
+				},
+				beforeFunc: func(t *testing.T, ctx context.Context, s Server, n service.NGT, test test) {
+					t.Helper()
+					var err error
+					if ir, err = request.GenMultiInsertReq(request.Float, vector.Gaussian, insertCnt, dim, defaultInsertConfig); err != nil {
+						t.Error(err)
+					}
+					if _, err := s.MultiInsert(ctx, ir); err != nil {
+						t.Error(err)
+					}
+				},
```

Comment:
[ask] 

When `beforeFunc` fails, is Error ok? 🤔 
I ask because I was wondering if we basically need to run tests on the target function when `beforeFunc` fails!

Reply:
`t.Error` fails the test so I think it is ok to use `t.Error`.
`t.Fatal` also fails the case, but it stops executing the test cases afterward, so I decided to use `t.Error` here.

Reply:
OK, I got it! Thank you!